doc/guides/design_guidelines.rst

```diff
@@ -0,0 +1,77 @@
+.. _design_guidelines:
+
+API Design Guidelines
+#####################
+
+Zephyr development and evolution is a group effort, and to simplify
+maintenance and enhancements there are some general policies that should
+be followed when developing a new capability or interface.
+
+Using Callbacks
+***************
+
+Many APIs involve passing a callback as a parameter or as a member of a
+configuration structure.  The following policies should be followed when
+specifying the signature of a callback:
+
+* The first parameter should be a pointer to the object most closely
+  associated with the callback.  In the case of device drivers this
+  would be ``struct device *dev``.  For library functions it may be a
+  pointer to another object that was referenced when the callback was
+  provided.
+
+* The next parameter(s) should be additional information specific to the
+  callback invocation, such as a channel identifier, new status value,
+  and/or a message pointer followed by the message length.
+
+* The final parameter should be a ``void *user_data`` pointer carrying
+  context that allows a shared callback function to locate additional
+  material necessary to process the callback.
+
+An exception to providing ``user_data`` as the last parameter may be
+allowed when the callback itself was provided through a structure that
+will be embedded in another structure.  An example of such a case is
+:c:type:`gpio_callback`, normally defined within a data structure
+specific to the code that also defines the callback function.  In those
+cases further context can accessed by the callback indirectly by
+:c:macro:`CONTAINER_OF`.
+
+Conditional Data and APIs
+*************************
+
+APIs and libraries may provide features that are expensive in RAM or
+code size but are optional in the sense that some applications can be
+implemented without them.  Examples of such feature include
+:option:`capturing a timestamp <CONFIG_CAN_RX_TIMESTAMP>` or
+:option:`providing an alternative interface <CONFIG_SPI_ASYNC>`.  The
+developer in coordination with the community must determine whether
+enabling the features is to be controllable through a Kconfig option.
+
+In the case where a feature is determined to be optional the following
+practices should be followed.
+
+First, any data that is accessed only when the feature is enabled should
+be conditionally included via ``ifdef CONFIG_FOO`` in the structure or
+union declaration.  This reduces memory use for applications that don't
+need the capability.
+
+Second, function declarations that are available only when the option is
+available may be handled in the either of these ways:
```

Comment:
Why have two recommended practices for this? How does one choose the best one?

Reply:
Both approaches are present in existing headers.  The rationale for preferring one over another is in the following paragraph, but there is no strong recommendation.  

I'm ambivalent about having both.  Removing the first (unconditional declarations+documentation) would be OK and maybe that's what should happen.  Then there should be a task to go through the headers and remove the unnecessary exclusions, and update the per-function documentation to explicitly note the dependency, to make the codebase consistent with the recommendations.

Reply:
I think we should try to pick only one (and I agree it should be unconditional declarations and documentation). Any APIs currently in violation of this should then be updated (can be done in separate PRs) to match the new recommendation.

Reply:
OK, we'll go that way.